src/pg_query_summary.c

```diff
@@ -0,0 +1,780 @@
+
+// Given a Node (node) and WalkState (state), wraps node in a RangeVarWithContext
+// and adds it to the state->range_vars list.
+static void add_range_var_with_context(Node* node, ContextType context, WalkState* state) {
+	if (!node)
+		return;
+
+	switch (node->type) {
+		case T_JoinExpr:
+		{
+			JoinExpr* je = castNode(JoinExpr, node);
+			add_range_var_with_context(je->larg, context, state);
+			add_range_var_with_context(je->rarg, context, state);
+			// je->quals is handled by the regular tree walk.
+			break;
+		}
+
+		case T_RangeTableSample:
+		{
+			RangeTableSample* rts = castNode(RangeTableSample, node);
+			add_range_var_with_context(rts->relation, context, state);
+			break;
+		}
+
+		case T_JsonTable:
+		case T_RangeTableFunc:
+		case T_RangeFunction:
+		case T_RangeSubselect:
+			// These are handled by the tree walker.
+			break;
+
+		case T_RangeVar: {
+			RangeVarWithContext* rvwc = palloc(sizeof(RangeVarWithContext));
+			rvwc->node = castNode(RangeVar, node);
+			rvwc->context = context;
+
+			state->range_vars = lappend(state->range_vars, rvwc);
+			break;
+		}
+
+		default:
+			elog(ERROR, "unexpected node type: %d", nodeTag(node));
+	}
+}
+
+// Given a List*-of-Nodes (clause) and WalkState (state), wrap them all
+// in RangeVarWithContext structs and add them to the state->range_vars list.
+static void add_range_vars_with_context(List *clause, ContextType context, WalkState *state) {
+	if (!clause)
+		return;
+
+	ListCell *lc = NULL;
+	foreach(lc, clause) {
+		add_range_var_with_context(lfirst(lc), context, state);
+	}
+}
+
+static void add_function(List* funcname, ContextType context, WalkState* state){
+	SummaryFunction* func = palloc(sizeof(SummaryFunction));
+
+	func->name = NameListToString(funcname);
+
+	if (list_length(funcname) == 3)
+		func->schema_name = lsecond_node(String, funcname)->sval;
+	else if (list_length(funcname) == 2)
+		func->schema_name = linitial_node(String, funcname)->sval;
+	else
+		func->schema_name = NULL;
+
+	func->function_name = llast_node(String, funcname)->sval;
+
+	func->context = context;
+
+	state->functions = lappend(state->functions, func);
+}
+
+static void add_table(WalkState* state, ContextType context, RangeVar* rv) {
+	if (rv->relname != NULL && cte_names_include(state, rv->relname))
+		return;
+
+	SummaryTable* table = palloc(sizeof(SummaryTable));
+	table->name = range_var_to_name(rv);
+	table->schema_name = pstrdup(rv->schemaname ? rv->schemaname : "");
+	table->table_name = pstrdup(rv->relname ? rv->relname : "");
+	table->context = context;
+
+	state->tables = lappend(state->tables, table);
+}
+
+bool pg_query_summary_walk_impl(Node* node, WalkState* state) {
+	if (node == NULL)
+		return false;
+
+	// NOTE: If `raw_expression_tree_walker()` doesn't handle a particular
+	// node type, you will need to `return false` early to avoid passing it
+	// to the tree walker.
+	// If you do this, be sure to update `pg_query_summary_stmt_walk_impl()`.
+
+	switch (nodeTag(node)) {
```

Comment:
There are certain utility statements not yet handled here, for example `AlterExtensionStmt`. Because the treewalker doesn't support those, it would cause an error - as discussed in `pg_query_summary_statement.c` maybe easiest to deal with by introducing a special function for checking whether a node type is not supported by the raw expression tree walker (and then returning false before we call the treewalker at the end of this function).

This will also let us consistently use `break;` in this switch statement without having to `return false`.